File: 2022/day01.hs
-- Advent of Code 2022, day 1
--   Venanzio Capretta

module Main where

import System.Environment
import Data.List
import Data.Char

import Control.Applicative
import qualified Data.Map as M

import FunParser
import AoCTools

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  input <- readFile fileName
  let xs = parseAll pInput input
  putStrLn ("Part 1: " ++ show (part1 xs))
  putStrLn ("Part 2: " ++ show (part2 xs))

-- Parsing the input

pData :: Parser [Int]
pData = many integer

pInput :: Parser [[Int]]
pInput = blocks pData

-- Part 1

part1 :: [[Int]] -> Int
part1 xs = maximum (map sum xs)

-- Part 2

part2 :: [[Int]] -> Int
part2 = sum . take 3 . reverse . sort . map sum

File: 2022/python/day01.py
# Advent of Code 2022, day 1
#   Venanzio Capretta

print("Advent of Code 2022, Day 1")

f = open("../input")
input = f.read().strip()
f.close()

# Parsing the input

# split a string in blocks separated by empty lines
def blocks(s):
  return (s.split('\n\n'))

# list of calories for one elf
def elfcal(s):
  return [int(x) for x in s.split()]

# calories for each food item for each elf
foodcal = [elfcal(s) for s in blocks(input)]

# Part 1

print("Part 1: ")

calories = [sum(xs) for xs in foodcal]

print(max(calories))

# Part 2

print("Part 2: ")
  
calories.sort(reverse=True)

print(sum(calories[0:3]))
